### PR TITLE
[DataFrame] Update architecture to be more flexible and performant

### PR DESCRIPTION
**Pull request by devin-petersohn**
## What do these changes do?

Changed the underlying architecture to be partitioned based on blocks and allow for data to be accessed by either row or column. Also updated many method implementations that needed minor changes to be correct in the new architecture.

### CODE REVIEW

File: python/ray/dataframe/dataframe.py [hunk 1]
```diff
@@ -9,123 +9,357 @@
...
+
+        # Sometimes we only get a single column or row, which is
+        # problematic for building blocks from the partitions, so we
+        # add whatever dimension we're missing from the input.
+        if self._blk_partitions.ndim != 2:
+            self._blk_partitions = np.expand_dims(self._blk_partitions,
+                                                  axis=axis ^ 1)
+
+        # Create the row and column index objects for using our partitioning.
+        self._row_lengths, self._row_index, \
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Split these into two.

File: python/ray/dataframe/dataframe.py [hunk 2]
```diff
@@ -9,123 +9,357 @@
...
+        """Set the _col_index for this DataFrame.
+
+        Args:
+            new__index: The new default index to set.
+        """
+        self._col_index_cache = new__index
+
+    _col_index = property(_get__col_index, _set__col_index)
+
+    def _compute_row_lengths(self):
```
Comment (pschafhalter) [inline python/ray/dataframe/dataframe.py]: Is this called anywhere? If not, remove.

File: python/ray/dataframe/dataframe.py [hunk 3]
```diff
@@ -9,123 +9,357 @@
...
+    def _compute_col_lengths(self):
         """Updates the stored lengths of DataFrame partions
         """
-        self._lengths = [_deploy_func.remote(_get_lengths, d)
-                         for d in self._df]
+        self._col_lengths = [_deploy_func.remote(lambda df: df.shape[1], d)
+                             for d in self._col_partitions]
 
-    def _get_lengths(self):
+    def _get_col_lengths(self):
```
Comment (pschafhalter) [inline python/ray/dataframe/dataframe.py]: This might not be called anywhere either.

File: python/ray/dataframe/dataframe.py [hunk 4]
```diff
@@ -217,38 +489,50 @@ def shape(self):
...
+
+        elif col_partitions is not None:
+            self._col_partitions = col_partitions
+
+        if row_partitions is not None or col_partitions is not None:
+            # At least one partition list is being updated, so recompute
+            # lengths and indices
+            self._row_lengths, self._row_index, \
+                self._col_lengths, self._col_index = \
+                _build_columns_and_index.remote(self._blk_partitions[:, 0],
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: TODO: split these into two.

File: python/ray/dataframe/dataframe.py [hunk 5]
```diff
@@ -912,81 +1105,112 @@ def drop(self, labels=None, axis=0, index=None, columns=None, level=None,
...
             raise ValueError("Need to specify at least one of 'labels', "
                              "'index' or 'columns'")
-        new_df = self
-        is_axis_zero = axis is None or axis == 0 or axis == 'index'\
-            or axis == 'rows'
-        try:
-            if (is_axis_zero and columns is None) or index is not None:
-                values = labels if labels is not None else index
+        obj = self.copy()
+
```
Comment (osalpekar) [inline python/ray/dataframe/dataframe.py]: TODO: clean up if-else block so code isn't duplicated

File: python/ray/dataframe/dataframe.py [hunk 6]
```diff
@@ -1162,79 +1401,149 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
...
-                    for i in range(len(self._df) - 1)
-                ]
+        if axis == 0:
+            parts = new_obj._col_partitions
+            idx_obj = new_obj._col_index
+        else:
+            parts = new_obj._row_partitions
+            idx_obj = new_obj._row_index
+
+        if isinstance(value, (pd.Series, dict)):
```
Comment (osalpekar) [inline python/ray/dataframe/dataframe.py]: raise NotImplementedError if value is a dataframe

File: python/ray/dataframe/dataframe.py [hunk 7]
```diff
@@ -1162,79 +1401,149 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
...
         else:
-            return new_df
+            if axis == 0:
+                new_obj._update_inplace(col_partitions=new_parts,
+                                        columns=self.columns)
+            else:
+                new_obj._update_inplace(row_partitions=new_parts,
+                                        columns=self.columns)
+            return new_obj
+        #
```
Comment (osalpekar) [inline python/ray/dataframe/dataframe.py]: TODO: delete commented out code below

File: python/ray/dataframe/dataframe.py [hunk 8]
```diff
@@ -1252,9 +1561,8 @@ def first_valid_index(self):
         Returns:
             scalar: type of index
         """
-        idx = self._index
-        if (idx is not None):
-            return idx.first_valid_index()
+        if self._row_index is not None:
+            return self._row_index.first_valid_index()
```
Comment (osalpekar) [inline python/ray/dataframe/dataframe.py]: TODO: fix `first_valid_index()`

File: python/ray/dataframe/dataframe.py [hunk 9]
```diff
@@ -1450,44 +1743,46 @@ def insert(self, loc, column, value, allow_duplicates=False):
             value (int, Series, or array-like): The values to insert.
             allow_duplicates (bool): Whether to allow duplicate column names.
         """
-        try:
-            len(value)
-        except TypeError:
+        # TODO: not working after _col_index rewrite
```
Comment (osalpekar) [inline python/ray/dataframe/dataframe.py]: remove TODO

File: python/ray/dataframe/dataframe.py [hunk 10]
```diff
@@ -1450,44 +1743,46 @@ def insert(self, loc, column, value, allow_duplicates=False):
             value (int, Series, or array-like): The values to insert.
             allow_duplicates (bool): Whether to allow duplicate column names.
         """
-        try:
-            len(value)
-        except TypeError:
+        # TODO: not working after _col_index rewrite 
+        if not is_list_like(value):
             value = [value for _ in range(len(self.index))]
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: Use `np.full` instead

File: python/ray/dataframe/dataframe.py [hunk 11]
```diff
@@ -1610,9 +1914,8 @@ def last_valid_index(self):
         Returns:
             scalar: type of index
         """
-        idx = self._index
-        if (idx is not None):
-            return idx.last_valid_index()
+        if self._row_index is not None:
```
Comment (osalpekar) [inline python/ray/dataframe/dataframe.py]: TODO: incorrect implementation

File: python/ray/dataframe/dataframe.py [hunk 12]
```diff
@@ -1911,51 +2201,56 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
...
 
     def query(self, expr, inplace=False, **kwargs):
         """Queries the Dataframe with a boolean expression
 
         Returns:
             A new DataFrame if inplace=False
         """
-        new_dfs = [_deploy_func.remote(lambda df: df.query(expr, **kwargs),
-                                       part) for part in self._df]
+        columns = self.columns
```
Comment (osalpekar) [inline python/ray/dataframe/dataframe.py]: TODO: does not work with local variables

File: python/ray/dataframe/dataframe.py [hunk 13]
```diff
@@ -1994,6 +2289,7 @@ def reindex_like(self, other, method=None, copy=True, limit=None,
 
     def rename(self, mapper=None, index=None, columns=None, axis=None,
                copy=True, inplace=False, level=None):
+        return # TODO: Fix this
```
Comment (osalpekar) [inline python/ray/dataframe/dataframe.py]: TODO: incorrect implementation

File: python/ray/dataframe/dataframe.py [hunk 14]
```diff
@@ -2138,7 +2435,10 @@ def _maybe_casted_values(index, labels=None):
                             values, mask, np.nan)
             return values
 
-        _, new_index = _compute_length_and_index.remote(new_obj._df)
+        _, new_index , _, _= \
```
Comment (osalpekar) [inline python/ray/dataframe/dataframe.py]: Split up  `_build_columns_and_index()`

File: python/ray/dataframe/dataframe.py [hunk 15]
```diff
@@ -3075,5 +3468,30 @@ def iloc(self):
         We currently support: single label, list array, slice object
         We do not support: boolean array, callable
         """
-        from .indexing import _iLoc_Indexer
-        return _iLoc_Indexer(self)
+        raise NotImplementedError(
+            "To contribute to Pandas on Ray, please visit "
+            "github.com/ray-project/ray.")
+
+    def _get_col_locations(self, col):
```
Comment (osalpekar) [inline python/ray/dataframe/dataframe.py]: Check if below 2 functions are used anywhere

File: python/ray/dataframe/test/test_dataframe.py [hunk 16]
```diff
@@ -1127,7 +1143,7 @@ def test_fillna():
     test_fillna_dtype_conversion()
     test_fillna_skip_certain_blocks()
     test_fillna_dict_series()
-    test_fillna_dataframe()
+    # test_fillna_dataframe()
```
Comment (osalpekar) [inline python/ray/dataframe/test/test_dataframe.py]: `pytest.raises(NotimplementedError)`

File: python/ray/dataframe/dataframe.py (hunk 6 above)
Comment (devin-petersohn) [reply python/ray/dataframe/dataframe.py]: Done above.

File: python/ray/dataframe/dataframe.py (hunk 11 above)
Comment (devin-petersohn) [reply python/ray/dataframe/dataframe.py]: I double checked these, they are correct.

File: python/ray/dataframe/dataframe.py [hunk 17]
```diff
@@ -9,123 +9,338 @@
...
+
+    def _set_row_partitions(self, new_row_partitions):
+        self._blk_partitions = \
+            _create_blk_partitions(new_row_partitions, axis=0,
+                                   length=len(self.columns))
+
+    _row_partitions = property(_get_row_partitions, _set_row_partitions)
+
+    def _get_col_partitions(self):
+        @ray.remote
```
Comment (robertnishihara) [inline python/ray/dataframe/dataframe.py]: avoid defining remote function here

File: python/ray/dataframe/dataframe.py [hunk 18]
```diff
@@ -401,7 +661,12 @@ def isnull(self):
             True: cell contains null.
             False: otherwise.
         """
-        return self._map_partitions(lambda df: df.isnull)
+        new_blk_partitions = np.array([_map_partitions(
```
Comment (robertnishihara) [inline python/ray/dataframe/dataframe.py]: blk -> block everywhere would be preferable

File: python/ray/dataframe/test/test_dataframe.py [hunk 19]
```diff
@@ -1143,6 +1161,8 @@ def test_fillna_sanity(num_partitions=2):
 
     zero_filled = test_data.tsframe.fillna(0)
     ray_df = from_pandas(test_data.tsframe, num_partitions).fillna(0)
+    print(ray_df)
+    print(zero_filled)
```
Comment (robertnishihara) [inline python/ray/dataframe/test/test_dataframe.py]: remove prints

File: python/ray/dataframe/dataframe.py [hunk 20]
```diff
@@ -9,123 +9,338 @@
...
-        self._df = df
-        self.columns = columns
-
-        # this _index object is a pd.DataFrame
-        # and we use that DataFrame's Index to index the rows.
-        self._lengths, self._index = _compute_length_and_index.remote(self._df)
-
-        if index is not None:
-            self.index = index
+            dtype : Data type to force. Only a single dtype is allowed.
```
Comment (robertnishihara) [inline python/ray/dataframe/dataframe.py]: `dtype:`

File: python/ray/dataframe/dataframe.py [hunk 21]
```diff
@@ -498,39 +742,25 @@ def all(self, axis=None, bool_only=None, skipna=None, level=None,
...
     def any(self, axis=None, bool_only=None, skipna=None, level=None,
             **kwargs):
-        """Return whether all elements are True over requested axis
+        """Return whether any elements are True over requested axis
 
         Note:
-            If axis=None or axis=0, this call applies df.all(axis=1)
-            to the transpose of df.
+            If axis=None or axis=0, this call applies on the column partitions,
+            otherwise operates on row partitions
```
Comment (robertnishihara) [inline python/ray/dataframe/dataframe.py]: indent by 4 spaces

File: python/ray/dataframe/dataframe.py [hunk 22]
```diff
@@ -217,38 +470,49 @@ def shape(self):
...
+            columns (pd.Index):
+                Index of the column dimension to replace existing columns
+            index (pd.Index):
+                Index of the row dimension to replace existing index
 
-    def _update_inplace(self, df=None, columns=None, index=None):
-        """Updates the current DataFrame inplace
+        Note:
+            If `columns` or `index` are not supplied, they will revert to
+            default columns or index respectively, as this function does not
```
Comment (robertnishihara) [inline python/ray/dataframe/dataframe.py]: indent by 4 spaces

**Review by robertnishihara — APPROVED** [review]
Thanks for addressing the comments. Currently failing linting, but looks good to me.

**Pull request merged**